sonar-kotlin-checks/src/main/java/org/sonarsource/kotlin/checks/VarShouldBeValCheck.kt

```diff
@@ -0,0 +1,107 @@
+import org.jetbrains.kotlin.psi.psiUtil.collectDescendantsOfType
+import org.jetbrains.kotlin.psi.psiUtil.findDescendantOfType
+import org.jetbrains.kotlin.psi.psiUtil.getParentOfType
+import org.jetbrains.kotlin.resolve.BindingContext
+import org.jetbrains.kotlin.resolve.BindingContext.REFERENCE_TARGET
+import org.jetbrains.kotlin.resolve.DescriptorToSourceUtils
+import org.sonar.check.Rule
+import org.sonarsource.kotlin.api.checks.AbstractCheck
+import org.sonarsource.kotlin.api.frontend.KotlinFileContext
+
+@Rule(key = "S3353")
+class VarShouldBeValCheck : AbstractCheck() {
+    override fun visitProperty(property: KtProperty, context: KotlinFileContext) {
+        if (property.isLocal && property.isVar) {
+            val bindingContext = context.bindingContext
+            val block = property.getParentOfType<KtBlockExpression>(false)!!
+            val assignments = collectAssignmentTo(block, property, bindingContext)
+
+            if (assignments.isEmpty()) {
+                context.reportIssue(property, """Replace the keyword `var` with `val`.""")
+            }
+        }
+    }
+
+    private fun collectAssignmentTo(block: PsiElement, variable: KtProperty, bindingContext: BindingContext): List<KtOperationExpression> {
+        KtDotQualifiedExpression
+        return block
+            .collectDescendantsOfType<KtOperationExpression> { expr -> isAssignment(expr) }
+            .filter { assignment ->
+                when (assignment) {
+                    is KtBinaryExpression -> isReferencingVar(variable, getReference(assignment), bindingContext) ?: false
+                    is KtUnaryExpression -> isReferencingVar(variable, getReference(assignment), bindingContext)
+                    // I don't know how to cover it
+                    else -> false
+                }
+            }
+    }
+
+    private fun getReference(operator: KtUnaryExpression): KtNameReferenceExpression {
+        val expr = operator.baseExpression!!
+        return if(expr is KtNameReferenceExpression){
+            expr
+        }else{
+            expr.findDescendantOfType<KtNameReferenceExpression>()!!
+        }
+    }
+
+
+    private fun isAssignment(expr: KtOperationExpression): Boolean {
+        return when (expr.operationReference.getReferencedNameElementType()) {
+            KtTokens.EQ -> true
+            KtTokens.PLUSEQ -> true
+            KtTokens.MINUSEQ -> true
+            KtTokens.MULTEQ -> true
+            KtTokens.DIVEQ -> true
+            KtTokens.PERCEQ -> true
+            KtTokens.PLUSPLUS -> true
+            KtTokens.MINUSMINUS -> true
+            else -> false
+        }
+    }
+
+    private fun getReference(expr: KtBinaryExpression): KtNameReferenceExpression {
+        return expr.left!!.findDescendantOfType<KtNameReferenceExpression>()!!
+    }
+
+    private fun isReferencingVar(variable: KtProperty, reference: KtNameReferenceExpression, bindingContext: BindingContext): Boolean {
```

Comment:
An idea to make the code more clean and consistent is to split the logic from the two `getReferene, the `isAssignment, and `isReferencingVar` to explicitly cover the case of `KtUnaryExpression` and `KtBinaryExpression`.
For instance having an `isAssignmentToVar` that checks that it is a `KtBinaryExpression` with the types EQ, PLUSEQ, ..., PERCEQ, and the appropriate reference is being assigned.
Then, an `isUnaryModificationToVar` checks that it is a `KtUnaryExpression` with the types PLUSPLUS and MINUSMINUS, and the appropriate reference is used.
Let's discuss it in chat eventually